google-cloud-container/lib/google/cloud/container/v1/cluster_manager_client.rb

```diff
@@ -317,13 +393,19 @@ def initialize \
           # Lists all clusters owned by a project in either the specified zone or all
           # zones.
           #
+          # @param parent [String]
+          #   The parent (project and location) where the clusters will be listed.
+          #   Specified in the format 'projects/*/locations/*'.
+          #   Location "-" matches all zones and all regions.
           # @param project_id [String]
```

Comment:
These newly deprecated arguments are changed from positional to named. That, and the addition of a _new_ positional argument, means that this is a breaking change. Users have to change their code when they upgrade. Since this is a breaking change, why even support these deprecated arguments at all?

Reply:
Oh dear. What we're seeing is a change to the way these calls specify the parent resource (as a single `parent` path parameter rather than separate `project` and `zone` parameters). At the API level, `parent` is now present and preferred, while `project` and `zone` are deprecated (but are still supported to maintain backward compatibility). Unfortunately, codegen always expresses the _preferred_ "parent resource" parameters as positional, so when they changed, this breaking change occurred.

Codegen will continue to generate the now-deprecated project and zone parameters because they still exist. I could synth them away, but that seems really brittle, and I'm not sure we really want to do that. Technically, those parameters should still be available because backward compatibility (even though it doesn't actually work that way because the two-part parent got replaced with one-part).

Reply:
In an ideal world, I'd like to see us not have any positional arguments at all, except for "the resource" (e.g. for "create" operations). In particular, the "parent" is a somewhat arbitrary concept that makes some sense in a REST context, but this is a gRPC API, not a REST API, nor are we presenting it as REST in the Ruby client. (And its arbitrariness is well-illustrated by the fact that the service changed what they thought the parent is.) And so "parents" should not be treated specially as positional arguments.

Unfortunately, that is not how GAPIC currently works. It is still trying to present things as REST-like.

I think when we approach the microgenerators, we need to keep this in mind. But for the current situation, the change is too big to do for the existing GAPIC generator, and too big to synth away, nor can we block this client while we wait for the microgenerator. So I'm inclined to proceed with the current code, knowing that it's a breaking change, and knowing that we're going to break it again. It's possible that when we move to the microgenerator, we can introduce some compatibility code that checks argument types.